src/zarr/abc/store.py

```diff
@@ -1,31 +1,58 @@
 from abc import ABC, abstractmethod
 from collections.abc import AsyncGenerator
-from typing import Protocol, runtime_checkable
+from typing import NamedTuple, Protocol, runtime_checkable
+
+from typing_extensions import Self
 
 from zarr.buffer import Buffer, BufferPrototype
-from zarr.common import BytesLike, OpenMode
+from zarr.common import AccessModeLiteral, BytesLike
+
+
+class AccessMode(NamedTuple):
+    readonly: bool
+    overwrite: bool
+    create: bool
+    update: bool
+
+    @classmethod
+    def from_literal(cls, mode: AccessModeLiteral) -> Self:
+        if mode in ("r", "r+", "a", "w", "w-"):
+            return cls(
+                readonly=mode == "r",
+                overwrite=mode == "w",
+                create=mode in ("a", "w", "w-"),
+                update=mode in ("r+", "a"),
+            )
+        raise ValueError("mode must be one of 'r', 'r+', 'w', 'w-', 'a'")
 
 
 class Store(ABC):
-    _mode: OpenMode
+    _mode: AccessMode
+
+    def __init__(self, mode: AccessModeLiteral = "r"):
+        self._mode = AccessMode.from_literal(mode)
 
-    def __init__(self, mode: OpenMode = "r"):
-        if mode not in ("r", "r+", "w", "w-", "a"):
-            raise ValueError("mode must be one of 'r', 'r+', 'w', 'w-', 'a'")
-        self._mode = mode
+        if self._exists():
+            if self.mode.update or self.mode.readonly:
+                pass
+            elif self.mode.overwrite:
+                self.clear()
+            else:
+                raise FileExistsError("Store already exists")
+
+    @abstractmethod
+    def _exists(self) -> bool: ...
+
+    @abstractmethod
+    def clear(self) -> None: ...
```

Comment:
I feel that these should be async methods. That may mean we need a class method to open a store (instead of using __init__.py).

Reply:
Something like:

```suggestion
   @classmethod
   @abstractmethod
    async def open(cls, mode: AccessModeLiteral = "r"):
        if await self.exists():
            ...
        return cls(mode=mode)

    @abstractmethod
    async def exists(self) -> bool: ...

    @abstractmethod
    async def clear(self) -> None: ...
```

Reply:
true! `async open` instead of `__init__ `seems to be the right way to go